### ISSUE TITLE
IntelliSense loses track of types of variables

### ISSUE THREAD

**Issue by Peter-Juhasz** [dotnet/roslyn#75652]
**Version Used**: 17.12 P4

**Steps to Reproduce**:

https://github.com/user-attachments/assets/4fde08d1-5ae8-4233-aa00-d8eda1c0df4b

**Diagnostic Id**:

**Expected Behavior**:

The type of `dataSource` should be clear, because:
- `dataSourceFactory.GetDataSourceAsync` returns a `ValueTask<IDataSource>` (awaited)
- it has no other overloads, in fact the only method of that interface (`GetDocumentAsync` has no overloads either)
- neither the syntax tree is broken which could invalidate the declarations earlier
- no other symbol named `dataSource` outside of this context
- none of the related types are declared in the enclosing scope

**Actual Behavior**:

Quick Info can't infer the type of `dataSource`, code completion doesn't work either.

Diagnostics don't report it as an individual error, it reports only that `GetDocumentAsync` is missing an argument - which is correct, and also means that some underlying parts could still infer the type correctly.

This must be an edge case, because I encounter this error rarely.

**Comment by CyrusNajmabadi** [comment]
I don't repro this.  @Peter-Juhasz the gifs are nice.  but we really need links to your code so we can attempt to realistically see what's oing on.

**Comment by Peter-Juhasz** [comment]
(off) @CyrusNajmabadi is there any way to help diagnose these issues which doesn't involve sending a full memory snapshot or full source code repository? In some cases, they only repro in a large project, even if I copy all related types to a small project (which is time consuming as well on both ends). This case may be even more tricky, because this doesn't look like an unhandled exception, but more like a logical error. Any Roslyn specific logs, which could help in any case to supplement the recordings?

**Comment by Peter-Juhasz** [comment]
@CyrusNajmabadi I was able to create a small repro! (+ `System.CommandLine` needed)

https://github.com/user-attachments/assets/191aa3b1-77b3-4e8a-9b3a-fa013ef470ea

```cs
using System.CommandLine;

namespace App.Client.Console;

public static partial class Commands
{
    public static void AddGetDocumentCommand(this Command parent, IDataSourceFactory dataSourceFactory)
    {
        var command = new Command("get", "Get a document from a data source.")
        {
        };

        command.SetHandler(async (context) =>
        {
            var cancellationToken = context.GetCancellationToken();

            var dataSourceId = "";
            var dataSource = await dataSourceFactory.GetDataSourceAsync(dataSourceId, cancellationToken);

            var documentId = "";

            await dataSource.GetDocumentAsync( cancellationToken);
        });
    }
}

public interface IDataSourceFactory
{
    ValueTask<IDataSource?> GetDataSourceAsync(string name, CancellationToken cancellationToken);
}

public interface IDataSource
{
    Task GetDocumentAsync(string id, CancellationToken cancellationToken);
}
```

If I remove the argument from the `GetDocumentAsync` call:
 - it loses track of both `dataSource`
 - and `context` as well, although they should be clear,
 - and also, it resolve `SetHandler` to the `Action<InvocationContext>` overload (instead of `Func<InvocationContext, Task>`), although the structure of the code has not changed, only an argument of many was removed

**Comment by CyrusNajmabadi** [comment]
Assigning to compiler.  This appears to be a case of poor error recovery .  @dotnet/roslyn-compiler I don't see why error recovery is falling down so badly here:

![Image](https://github.com/user-attachments/assets/855da367-2079-4e40-b310-217a675bc50a)

There is only one GetDataSourceAsync method off of this type, and it returns a `Task<IDataSource?>`.  `await`ing that should clearly produce an `IDataSource`.  Can you guys ptal?

**Comment by jaredpar** [comment]
When building the sample you get the following error:

> C:\Users\jaredpar\temp\console\Program.cs(22,30): error CS7036: There is no argument given that corresponds to the required parameter 'cancellationToken' of 'IDataSource.GetDocumentAsync(string, CancellationToken)' [C:\Users\jaredpar\temp\console\console.csproj]

Compiler seems to be detecting the overload just fine here. Is it just giving different results via the API?
